src/devices/Pcx857x/tests/Pcx857xTest.cs

```diff
@@ -0,0 +1,177 @@
+            public override byte ReadByte() => DeviceMock.ReadByte();
+        }
+
+        /// <summary>
+        /// Mock the behavior of the chip
+        /// </summary>
+        public class Pcx857xChipMock
+        {
+            private int _ports;
+            private byte[] _registers;
+            private byte[] _lastReadBuffer;
+            private byte[] _lastWriteBuffer;
+
+            public Pcx857xChipMock(int ports)
+            {
+                _ports = ports;
+                _registers = new byte[ports];
+            }
+
+            public Span<byte> Registers => _registers;
+
+            // Can't coalesce here https://github.com/dotnet/roslyn/issues/29927
+            public ReadOnlySpan<byte> LastReadBuffer => _lastReadBuffer == null ? ReadOnlySpan<byte>.Empty : _lastReadBuffer;
+            public ReadOnlySpan<byte> LastWriteBuffer => _lastWriteBuffer == null ? ReadOnlySpan<byte>.Empty : _lastWriteBuffer;
+
+            public void Read(Span<byte> buffer)
+            {
+                _lastReadBuffer = buffer.ToArray();
+
+                for (int i = 0; i < buffer.Length; i++)
+                {
+                    buffer[i] = _registers[i];
+                }
+            }
+
+            public void Write(Span<byte> data)
+            {
+                _lastWriteBuffer = data.ToArray();
+
+                for (int i = 0; i < data.Length; i++)
+                {
+                    _registers[i] = data[i];
+                }
+            }
+
+            public byte ReadByte()
+            {
+                Span<byte> buffer = stackalloc byte[1];
+                Read(buffer);
+                return buffer[0];
+            }
+
+            public void WriteByte(byte value)
+            {
+                Span<byte> buffer = stackalloc byte[] { value };
+                Write(buffer);
+            }
+        }
+
+
+        public class GpioControllerMock : IGpioController
+        {
+            private Dictionary<int, PinValue> _pinValues = new Dictionary<int, PinValue>();
+
+            public void Reset() => _pinValues = new Dictionary<int, PinValue>();
+
+            public void ClosePin(int pinNumber)
+            {
+            }
+
+            public void Dispose()
+            {
+            }
+
+            public void OpenPin(int pinNumber, PinMode mode)
+            {
+            }
+
+            public PinValue Read(int pinNumber)
+            {
+                if (_pinValues.TryGetValue(pinNumber, out PinValue value))
+                    return value;
+
+                return PinValue.Low;
+            }
+
+            public void Read(Span<PinValuePair> pinValues)
+            {
+                for (int i = 0; i < pinValues.Length; i++)
+                {
+                    int pin = pinValues[i].PinNumber;
+                    pinValues[i] = new PinValuePair(pin, Read(pin));
+                }
+            }
+
+            public void SetPinMode(int pinNumber, PinMode mode)
+            {
```

Comment:
If not expected to be called or expected to be called with specific values might be worth adding some assert here.

Reply:
We don't care what it gets called with yet. When we add tests that track that we'll add an implementation.  I'll also be rejiggering this a bit to use MOQ in a later PR.